BTCPayServer/Payments/PayJoin/PayJoinEndpointController.cs

```diff
@@ -28,374 +34,372 @@ public class PayJoinEndpointController : ControllerBase
         [EnableCors(CorsPolicies.All)]
         [MediaTypeConstraint("text/plain")]
         [RateLimitsFilter(ZoneLimits.PayJoin, Scope = RateLimitsScope.RemoteAddress)]
-        public async Task<IActionResult> Submit(string cryptoCode, string invoice)
+        public async Task<IActionResult> Submit(string cryptoCode)
         {
             var network = _btcPayNetworkProvider.GetNetwork<BTCPayNetwork>(cryptoCode);
             if (network == null)
             {
-                return UnprocessableEntity("Incorrect network");
+                return BadRequest(CreatePayjoinError(400, "invalid-network", "Incorrect network"));
             }
 
-            string rawBody;
-            using (StreamReader reader = new StreamReader(Request.Body, Encoding.UTF8))
+            if (Request.ContentLength is long length)
             {
-                rawBody = await reader.ReadToEndAsync();
+                if (length > 1_000_000)
+                    return this.StatusCode(413,
+                        CreatePayjoinError(413, "payload-too-large", "The transaction is too big to be processed"));
             }
-
-            if (string.IsNullOrEmpty(rawBody))
+            else
             {
-                return UnprocessableEntity("raw tx not provided");
+                return StatusCode(411,
+                    CreatePayjoinError(411, "missing-content-length",
+                        "The http header Content-Length should be filled"));
             }
 
-            PSBT psbt = null;
-            if (!Transaction.TryParse(rawBody, network.NBitcoinNetwork, out var transaction) &&
-                !PSBT.TryParse(rawBody, network.NBitcoinNetwork, out psbt))
+            string rawBody;
+            using (StreamReader reader = new StreamReader(Request.Body, Encoding.UTF8))
             {
-                return UnprocessableEntity("invalid raw transaction or psbt");
+                rawBody = (await reader.ReadToEndAsync()) ?? string.Empty;
             }
 
-            if (psbt != null)
+            Transaction originalTx = null;
+            FeeRate originalFeeRate = null;
+            int originalVSize = 0;
+            bool psbtFormat = true;
+            if (!PSBT.TryParse(rawBody, network.NBitcoinNetwork, out var psbt))
             {
-                try
-                {
-                    transaction = psbt.ExtractTransaction();
-                }
-                catch (Exception e)
+                psbtFormat = false;
+                if (!Transaction.TryParse(rawBody, network.NBitcoinNetwork, out var tx))
+                    return BadRequest(CreatePayjoinError(400, "invalid-format", "invalid transaction or psbt"));
+                psbt = PSBT.FromTransaction(tx, network.NBitcoinNetwork);
+                for (int i = 0; i < tx.Inputs.Count; i++)
                 {
-                    return UnprocessableEntity("invalid psbt");
+                    psbt.Inputs[i].FinalScriptSig = tx.Inputs[i].ScriptSig;
```

Comment:
I always found this weird. Why does a psbt created from a tx lose its inputs' signatures?

Reply:
The method is not well named. It should be `PSBT.FromGlobalTransaction`. Which make things clear. Global Transaction is the "template" of the transaction being expressed by the PSBT, without sigs.